### PR TITLE
[WIP]: Do not pick hostname which resolves to IP not present locally

### LINKED ISSUE #8211: Node chooses wrong hostname
We have a join configuration like this for Hazelcast 3.6.2:

``` java
config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
config.getNetworkConfig().getInterfaces().setInterfaces(interfaces).setEnabled(true);
config.getNetworkConfig().getJoin().getTcpIpConfig().setEnabled(true).setMembers(servers);
```

The String lists come out of our Spring Boot configuration.
For local development this settings work fine:

hazelcast:
  servers:
    - 127.0.0.1
  interfaces:
    - 127.0.0.1

In our production and test enviroment we tried to use this configuration:

hazelcast:
  servers:
    - appnode1
    - appnode2
  interfaces:
    - "_._._._"

Now every server thinks he is appnode1 because Hazelcast just resolves appnode1 and appnode2 and then matches the interfaces mask onto it. The first match wins. So every node chooses appnode1. 

If we disable the interfaces feature everything works as expected. Hazelcast enumerates every ip on the host and matches them against the resolved hostnames.

Shouldn't the whole process work a little bit different? e.g. Find out all IPs of the current host, match them against the interfaces filter strings and compare them to the ips resolved out of the hostnames to determine the nodes hostname?

### PR DESCRIPTION
**Pull request by jerrinot**
Fixes #8211 

I tried the simplest change possible. It needs more testing, especially around ipv6.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java [hunk 1]
```diff
@@ -198,6 +174,37 @@ private AddressDefinition pickInterfaceAddressDef() throws UnknownHostException,
         return interfaces;
     }
 
+    private Map<String, String> createAddressDomainMap(TcpIpConfig tcpIpConfig) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java]: this is mostly just extracted code from `getInterfaces()`. the only change are two `AddressUtil.matchAnyLocalInterface(addressHolder)` checks

File: hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java [hunk 2]
```diff
@@ -198,6 +174,37 @@ private AddressDefinition pickInterfaceAddressDef() throws UnknownHostException,
...
+            if (AddressUtil.isIpAddress(addressHolder)) {
+                // there may be a domain registered for this address
+                if (!addressDomainMap.containsKey(addressHolder) && AddressUtil.matchAnyLocalInterface(addressHolder)) {
+                    addressDomainMap.put(addressHolder, null);
+                }
+            } else {
+                try {
+                    Collection<String> addresses = resolveDomainNames(addressHolder);
+                    for (String address : addresses) {
+                        if (AddressUtil.matchAnyLocalInterface(address)) {
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java]: This doesn't seem efficient - it leads to too many calls of `matchersForAllLocalInterfaces` if there is more members (and they resolves to more IP adddresses).

**Comment by mmedenjak** [comment]
Superseded by https://github.com/hazelcast/hazelcast/pull/13884?

**Comment by mmedenjak** [comment]
Moving to 3.12 as it is not a blocker for the release and there's an alternative PR

**Comment by jerrinot** [comment]
closing in favour of https://github.com/hazelcast/hazelcast/pull/13884

**Pull request closed without merging**